### PR TITLE
Added fc + activation fuse pass (currently only gelu, sigmoid and tanh are supported)

### PR DESCRIPTION
**Pull request by jakpiase**
### PR types
New features

### PR changes
Others

### Describe
Added oneDNN fuse pass for FC + activation. Currently gelu, sigmoid and tanh are supported. 

Performance improvement is around 4,5% - benchmarked on BERT model using Intel(R) d on(R) Gold 6348H CPU @ 2.30GHz

image

I1218 22:56:33.140385 102746 helper.h:363] ====== threads: 1, thread id: 0 ======
I1218 22:56:33.140409 102746 helper.h:365] ====== batch size: 1, iterations: 4962, repetitions: 1 ======
I1218 22:56:33.140413 102746 helper.h:367] ====== batch latency: 3.02517ms, number of samples: 4962, sample latency: 3.02517ms, fps: 330.56, data type: float ======

### CODE REVIEW

**Comment by jakpiase** [comment]
@jczaja Could you please take a look?

File: paddle/fluid/framework/ir/graph_pattern_detector.cc [hunk 1]
```diff
@@ -1017,6 +1017,26 @@ PDNode *patterns::FCMKLDNN::operator()(paddle::framework::ir::PDNode *x,
   return fc_out_var;
 }
 
+PDNode *patterns::FCActOneDNN::operator()(const std::string &act_type) {
+  auto *fc_x = pattern->NewNode(fc_in_repr())
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/graph_pattern_detector.cc]: FC's input is redundant in this pattern.

File: paddle/fluid/framework/ir/graph_pattern_detector.h [hunk 2]
```diff
@@ -552,6 +552,28 @@ struct FCMKLDNN : public PatternBase {
...
+//          Formula: act(fc(x))
+//          Op: fc + act
+struct FCActOneDNN : public PatternBase {
+  FCActOneDNN(PDPattern* pattern, const std::string& name_scope)
+      : PatternBase(pattern, name_scope, "fc_act_onednn") {}
+
+  PDNode* operator()(const std::string& act_type);
+
+  // declare operator node's name
+  PATTERN_DECL_NODE(fc_in);
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/graph_pattern_detector.h]: ditto

File: paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass_tester.cc [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+  EXPECT_THROW(RunPassAndAssert(&graph, "x", "act_y", removed_nodes_count),
+               paddle::platform::EnforceNotMet);
+}
+
+}  // namespace ir
+}  // namespace framework
+}  // namespace paddle
+
+USE_PASS(fc_act_fuse_pass);
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass_tester.cc]: Please, add a new line at the end of the file.

File: paddle/fluid/inference/api/paddle_pass_builder.cc [hunk 4]
```diff
@@ -206,7 +206,8 @@ void CpuPassStrategy::EnableMKLDNN() {
              "reshape_transpose_matmul_mkldnn_fuse_pass",  //
              "matmul_transpose_reshape_fuse_pass",         //
              // Disabled due to topology-dependent speed-up
-             // "fc_mkldnn_pass",
+             //"fc_mkldnn_pass",
+             //"fc_act_fuse_pass",
```
Comment (wojtuss) [inline paddle/fluid/inference/api/paddle_pass_builder.cc]: Should the pass be enabled conditionally somewhere?

File: paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass_tester.cc [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  CreateOp(&prog, "gelu", {{"Input", "fc_y"}}, {{"Out", "act_y"}}, false);
+
+  Graph graph(prog);
+  // No fusion in this attribute configuration
+  constexpr int removed_nodes_count = 0;
+
+  EXPECT_THROW(RunPassAndAssert(&graph, "x", "act_y", removed_nodes_count),
+               paddle::platform::EnforceNotMet);
+}
+
```
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass_tester.cc]: Please add:
*  _positive_ test cases and check correct fusing behaviour for all available activations {"gelu", "tanh", "gelu_erf", gelu_tanh"}. 
* checks for op attribute values correctness. (you're setting `activation_type` attribute in fuse pass)

File: paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass.cc [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+}  // namespace framework
+}  // namespace paddle
+
+REGISTER_PASS(fc_act_fuse_pass, paddle::framework::ir::FuseFCActOneDNNPass);
+REGISTER_PASS_CAPABILITY(fc_act_fuse_pass)
+    .AddCombination(
+        paddle::framework::compatible::OpVersionComparatorCombination()
+            .LE("fc", 0)
+            .LE("gelu", 0)
+            .LE("tanh", 0));
```
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass.cc]: Please add new line et EOF.

File: paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass_tester.cc (hunk 3 above)
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/fc_act_fuse_pass_tester.cc]: Please add new line at EOF.

**Comment by jczaja** [comment]
@jakpiase  Don't you need to modify https://github.com/PaddlePaddle/Paddle/blob/0cc42e34c68b16bee148a024cbf85bcb1eeabd91/paddle/fluid/operators/fc_op.cc#L82-L90 so that other activation than relu are allowed?

**Review by arogowie-intel — CHANGES_REQUESTED** [review]
You need also a functional test in python like [this one](https://github.com/PaddlePaddle/Paddle/blob/1af4cb63e8522d322570eb47c0c38edd1a9e45a1/python/paddle/fluid/tests/unittests/ir/inference/test_mkldnn_batch_norm_act_fuse_pass.py).

File: paddle/fluid/framework/ir/graph_pattern_detector.cc [hunk 7]
```diff
@@ -1017,6 +1017,26 @@ PDNode *patterns::FCMKLDNN::operator()(paddle::framework::ir::PDNode *x,
...
+                   ->AsInput()
+                   ->assert_is_op_input("fc", "Input");
+  auto *fc = pattern->NewNode(fc_repr())->assert_is_op("fc");
+  auto *fc_out = pattern->NewNode(fc_out_repr())
+                     ->assert_is_op_output("fc", "Out")
+                     ->assert_is_op_input(act_type);
+  auto *act =
```
Comment (arlesniak) [inline paddle/fluid/framework/ir/graph_pattern_detector.cc:1025]: Maybe after fuse you should remove intermediate orphaned nodes from graph.

**Comment by jakpiase** [comment]
@jczaja @wojtuss @arogowie-intel @arlesniak I have implemented all suggested changes. Could you all take a look?

**Review by jczaja — DISMISSED** [review]
LGTM

**Comment by jczaja** [comment]
@luotao1 Could you please start your review?

File: paddle/fluid/inference/api/paddle_pass_builder.cc (hunk 4 above)
Comment (luotao1) [reply paddle/fluid/inference/api/paddle_pass_builder.cc]: @jczaja Is this pass for native fc or fc_mkldnn? How about the name `fc_act_mkldnn_fuse_pass`?

**Review by wojtuss — COMMENTED** [review]
I agree with @luotao1, the pass name could be `fc_act_mkldnn_fuse_pass`.
Apart from that LGTM.

**Review by jczaja — DISMISSED** [review]
LGTM

File: paddle/fluid/inference/api/paddle_pass_builder.cc (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/inference/api/paddle_pass_builder.cc]: @luotao1 Name of pass was modified as suggested.

**Comment by luotao1** [comment]
@lidanqing-intel https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/2241079/job/3182633
```
2020-12-24 05:35:16 Total Test time (real) = 3892.54 sec
2020-12-24 05:35:16 The following tests FAILED:
2020-12-24 05:35:16 	830 - test_conv2d_int8_mkldnn_op (Failed)
```
test_conv2d_int8_mkldnn_op failed after merge https://github.com/PaddlePaddle/Paddle/pull/29528. I will revert it at first.

**Review by luotao1 — APPROVED** [review]

**Pull request merged**